src/sink/elasticsearch.rs

```diff
@@ -65,6 +66,54 @@ impl Elasticsearch {
             flush_interval: config.flush_interval,
         }
     }
+
+    fn ensure_indexed(&mut self,
+                      idx: String,
+                      doc: Payload)
+                      -> Result<(), elastic::error::Error> {
+        use elastic::error;
+        let get_res = self.client
+            .get_document::<Payload>(index(idx.clone()), id(doc.uuid.clone()))
+            .send();
+
+        match get_res {
+            // The doc was found: no need to index
+            Ok(GetResponse { source: Some(_), .. }) => {
+                report_telemetry("sinks.elasticsearch.success.doc_existed", 1.0);
+                Ok(())
+            }
+            // The index exists, but the doc wasn't found: map and index
+            Ok(_) => self.put_doc(idx.clone(), doc).map(|_| ()),
+            // No index: create it, then map and index
+            Err(error::Error(error::ErrorKind::Api(error::ApiError::IndexNotFound {
+                                                         ..
+                                                     }),
+                                      _)) => {
+                self.put_index(idx.clone()).and(self.put_doc(idx, doc)).map(|_| ())
```

Comment:
I'm not totally satisfied with the error handling here. You're correct, it's possible that two simultaneous writers could step on each other. I need to learn more about Elasticsearch's write model to learn what kind of error it is. 

If in fact this does result in an error, the losing write will get pushed to the back of the buffer and re-tried at a later date. Not really ideal if it can be helped. There should also be documentation in this module / around this concern.